feat: support insert multiple rows into a table using a single SQL insert statement

jingchen2222 opened the pull request:
* **Please check if the PR fulfills these requirements**

- The commit message follows our guidelines
- Tests for the changes have been added (for bug fixes / features)
- Docs have been added / updated (for bug fixes / features)


* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
#398 #391 
* **What is the new behavior (if this is a feature change)?**



* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)



* **Other information**:
Further things to do:
  - API Server support multi-rows insert statement @vagetablechicken 
  - Integration Test @wuyou10206

jingchen2222 committed:
feat: support insert multiple rows into a table using a single SQL insert statement.

Shouren commented on src/sdk/sql_cluster_router.cc:
```diff
@@ -261,7 +261,86 @@ std::shared_ptr<SQLInsertRow> SQLClusterRouter::GetInsertRow(const std::string&
+                                              dynamic_cast<::hybridse::node::ExprListNode*>(value),
+                                              &str_length));
+        if (!default_maps->back()) {
+            status->msg = "get default value map of row[" + std::to_string(i) + "] failed";
```
I think It would be better to contain the index info of row in the error message - e.g. "get default value map of row [xxx] ({index}/{total}) failed", but not sure the current format is good enough.

jingchen2222 replied:
Good idea! I will fix it. Thanks.

Shouren commented on src/sdk/sql_cluster_router.cc:
```diff
@@ -865,38 +944,58 @@ std::shared_ptr<hybridse::sdk::ResultSet> SQLClusterRouter::ExecuteSQLBatchReque
     }
-    if (!row->IsComplete()) {
-        status->msg = "insert value isn't complete";
+    if (default_maps.size() != str_lengths.size()) {
```
Perhaps this check should be put inside the ```GetMultiRowInsertInfo``` function, since they both come from that funciton.

jingchen2222 replied:
Fixed

Shouren commented on src/sdk/sql_cluster_router.cc:
```diff
@@ -865,38 +944,58 @@ std::shared_ptr<hybridse::sdk::ResultSet> SQLClusterRouter::ExecuteSQLBatchReque
+            return false;
+        }
+        std::vector<std::shared_ptr<::openmldb::catalog::TabletAccessor>> tablets;
+        bool ret = cluster_sdk_->GetTablet(db, table_info->name(), &tablets);
```
Perhaps ```GetTablet``` function should be outside the loop section, i don't know whether there is load-balance algorithm inside this function so we should call it before ```PutRow```.

jingchen2222 replied:
Actually, there isn't a dynamic load-balance algorithm here. So we can simply put `GetTablet` out of the loop. Just a bad design. I will fix it.

Shouren reviewed (changes requested):
Having a strong concern on the [duplicate call](https://github.com/4paradigm/OpenMLDB/pull/399/files#r706080907) to ```GetTablelet``` if there is no other reason to avoid the change.

jingchen2222 committed:
feat: resolve commnet
1. put  out of the loops
2. error message report success/total rows.
3. put default_maps size check into  method.

jingchen2222 committed:
fix: toydb dbms sdk test

jingchen2222 committed:
feat: add insert case for bool type insert

imotai reviewed (approved):
LGTM
